src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostAssignability.scala

```diff
@@ -122,14 +122,32 @@ trait GhostAssignability {
 
   /** conservative ghost separation assignment check */
   private[separation] def ghostAssignableToId(exprs: PExpression*)(lefts: PIdnNode*): Messages =
-    generalGhostAssignableTo(ghostExprResultTyping)(dfltGhostAssignableMsg(ghostIdClassification))(exprs: _*)(lefts: _*)
+    generalGhostAssignableTo(ghostExprResultTyping)(dfltGhostAssignableMsg[PIdnNode](wellGhostSeparated)(ghostIdClassification))(exprs: _*)(lefts: _*)
 
   /** conservative ghost separation assignment check */
-  private[separation] def ghostAssignableToParam(exprs: PExpression*)(lefts: PParameter*): Messages =
-    generalGhostAssignableTo(ghostExprResultTyping)(dfltGhostAssignableMsg(ghostParameterClassification))(exprs: _*)(lefts: _*)
+  private[separation] def ghostAssignableToParam(exprs: PExpression*)(lefts: PParameter*): Messages = {
+    generalGhostAssignableTo(ghostExprResultTyping)(dfltGhostAssignableMsg[PParameter](wellGhostSeparated)(ghostParameterClassification))(exprs: _*)(lefts: _*)
+  }
 
-  private def dfltGhostAssignableMsg[L <: PNode](ghost: L => Boolean): (Boolean, L) => Messages = {
-    case (g, l) => error(l, "ghost error: ghost cannot be assigned to non-ghost", g && !ghost(l))
+  /**
+    * default factory for producing messages that a ghost right-hand side cannot be assigned to a non-ghost left-hand side
+    * @param pre well-definedness condition for a left-hand side that must hold before `ghost` might be invoked
+    * @param ghost classifier whether a left-hand side is ghost
+    * @tparam L type of the left-hand sides
+    * @return
+    */
+  private def dfltGhostAssignableMsg[L <: PNode](pre: WellDefinedness[L])(ghost: L => Boolean): (Boolean, L) => Messages = {
+    case (g, l) =>
+      val lhsMsgs = pre(l).out
+      if (lhsMsgs.isEmpty) {
+        // `l` is well-defined
+        error(l, "ghost error: ghost cannot be assigned to non-ghost", g && !ghost(l))
+      } else {
+        // We return the well-definedness messages of 'l' even though these messages should already
+        // get reported when visiting 'l' in the AST.
+        // This ensures that we do not miss them and deduplication will remove duplicate messages
```

Comment:
I'm not sure I understand why this is necessary if we are just reporting messages that have been reported before. When could we "miss" these messages if we returned `noMessages` here instead?

Reply:
My rational is that if we ever change ghost separation to no longer iterate over all AST nodes (for whatever reason) and, thus, these messages do not get reported anymore when visiting the correspond `PParameter` node then returning here `noMessages` (aka everything looks good) despite not performing the actual checks seemed dangerous to me. If we are convinced that we would catch such a buggy code change in the future via our test suite then I'm happy to slightly simplify the code, check validity of `l` only (as opposed to collecting its messages), and return `noMessages` if it's invalid

Reply:
I've created issue #857 for this